app/channel.go

```diff
@@ -1121,6 +1121,15 @@ func (a *App) PatchChannelModerationsForChannel(c request.CTX, channel *model.Ch
 
 	cErr := a.forEachChannelMember(c, channel.Id, func(channelMember model.ChannelMember) error {
 		a.Srv().Store().Channel().InvalidateAllChannelMembersForUser(channelMember.UserId)
+
+		evt := model.NewWebSocketEvent(model.WebsocketEventChannelMemberUpdated, "", "", channelMember.UserId, nil, "")
```

Comment:
Passing the `userId` parameter to the `NewWebSocketEvent` function causes the [websocket broadcast to only be delivered to the that user](https://pkg.go.dev/github.com/mattermost/mattermost-server@v5.11.1+incompatible/model#WebsocketBroadcast). Is that your intended behaviour?

Reply:
In this case, we need to broadcast to the current members of the channel as those affected by this, for example, by not allowing the creation of a post. You wouldn't need that for channels you're not a part of yet, as you'll get the info when you join.

Reply:
ChannelMembers records do not change upon channel moderation updates. Roles records do change. Given that fact, I'm trying to understand why we need to broadcast each user in the channel their ChannelMember record?

Reply:
With the default settings, we do have that the user has `channel_user` role for this channel in Redux.

<img width="530" alt="image" src="https://user-images.githubusercontent.com/684680/211535488-0546563a-6304-4232-a17d-8dca89337a8e.png">

If we remove the create post, it will create a custom role for this channel that needs to be passed down.

<img width="956" alt="image" src="https://user-images.githubusercontent.com/684680/211535663-79e81e78-b223-4cf6-934d-e1b5d398cf7f.png">

We need to trigger this event so we can get the new role:

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/684680/211535959-c779a3f2-9687-438b-9fc8-6feb929a53e4.png">

Which will trigger this change in Redux:

<img width="1361" alt="image" src="https://user-images.githubusercontent.com/684680/211536990-6d01d399-ad6f-4eb9-a5de-1883048d090a.png">

Before this change, if you had an existing role and you changed some permissions without allowing everything, it worked as expected, as the user had a role already. If you went from `custom_role` -> `channel_user` -> `new_custom_role`, this information wouldn't be sent to the client.

The reason that only channel members need those updates is that each member should get the updated role for them. If, for example, there are 3 members, a sysadmin, a channel admin, and a channel user, they'll have different roles, so when the sysadmin changes the permissions, they'll need their own event.

Reply:
Got it. It's been a while, but you're reminded me that the front-end (of web/desktop, at least) is hiding the channel-scheme association by denormalizing the roles directly onto the channelmember.